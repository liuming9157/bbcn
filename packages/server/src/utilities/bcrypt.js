// TODO: REMOVE

const bcrypt = require("bcryptjs")
const env = require("../environment")

const SALT_ROUNDS = env.SALT_ROUNDS || 10

exports.hash = async data => {
  const salt = await bcrypt.genSalt(SALT_ROUNDS)
  const result = await bcrypt.hash(data, salt)
  return result
}

exports.compare = async (data, encrypted) =>
  await bcrypt.compare(data, encrypted)
