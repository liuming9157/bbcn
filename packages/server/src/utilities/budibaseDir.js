const { join } = require("./centralPath")
const { homedir } = require("os")
const env = require("../environment")
const { budibaseTempDir } = require("@budibase/backend-core/objectStore")

module.exports.budibaseAppsDir = function () {
  return env.BUDIBASE_DIR || join(homedir(), ".budibase")
}

module.exports.budibaseTempDir = budibaseTempDir
