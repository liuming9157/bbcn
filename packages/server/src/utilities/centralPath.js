const path = require("path")

// this simply runs all of our path join and resolve functions through
// a central location incase we need to add some protection to file paths

/**
 * Exactly the same as path.join
 * @param args Any number of string arguments to add to a path
 * @returns {string} The final path ready to use
 */
exports.join = function (...args) {
  return path.join(...args)
}

/**
 * Exactly the same as path.resolve
 * @param args Any number of string arguments to add to a path
 * @returns {string} The final path ready to use
 */
exports.resolve = function (...args) {
  return path.resolve(...args)
}
