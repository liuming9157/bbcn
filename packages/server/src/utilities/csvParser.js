const csv = require("csvtojson")
const { FieldTypes } = require("../constants")

const VALIDATORS = {
  [FieldTypes.STRING]: () => true,
  [FieldTypes.OPTIONS]: () => true,
  [FieldTypes.NUMBER]: attribute => {
    // allow not to be present
    if (!attribute) {
      return true
    }
    return !isNaN(Number(attribute))
  },
  [FieldTypes.DATETIME]: attribute => {
    // allow not to be present
    if (!attribute) {
      return true
    }
    return !isNaN(new Date(attribute).getTime())
  },
}

const PARSERS = {
  [FieldTypes.NUMBER]: attribute => {
    if (!attribute) {
      return attribute
    }
    return Number(attribute)
  },
  [FieldTypes.DATETIME]: attribute => {
    if (!attribute) {
      return attribute
    }
    return new Date(attribute).toISOString()
  },
}

function parse(csvString, parsers) {
  const result = csv().fromString(csvString)

  const schema = {}

  return new Promise((resolve, reject) => {
    result.on("header", headers => {
      for (let header of headers) {
        schema[header] = {
          type: parsers[header] ? parsers[header].type : "string",
          success: true,
        }
      }
    })
    result.subscribe(row => {
      // For each CSV row parse all the columns that need parsed
      for (let key of Object.keys(parsers)) {
        if (!schema[key] || schema[key].success) {
          // get the validator for the column type
          const validator = VALIDATORS[parsers[key].type]

          try {
            // allow null/undefined values
            schema[key].success = !row[key] || validator(row[key])
          } catch (err) {
            schema[key].success = false
          }
        }
      }
    })
    result.on("done", error => {
      if (error) {
        console.error(error)
        reject(error)
      }

      resolve(schema)
    })
  })
}

function updateSchema({ schema, existingTable }) {
  if (!schema) {
    return schema
  }
  const finalSchema = {}
  const schemaKeyMap = {}
  Object.keys(schema).forEach(key => (schemaKeyMap[key.toLowerCase()] = key))
  for (let [key, field] of Object.entries(existingTable.schema)) {
    const lcKey = key.toLowerCase()
    const foundKey = schemaKeyMap[lcKey]
    if (foundKey) {
      finalSchema[key] = schema[foundKey]
      finalSchema[key].type = field.type
    }
  }
  return finalSchema
}

async function transform({ schema, csvString, existingTable }) {
  const colParser = {}

  // make sure the table has all the columns required for import
  if (existingTable) {
    schema = updateSchema({ schema, existingTable })
  }

  for (let [key, field] of Object.entries(schema)) {
    // don't import data to auto columns
    if (!field.autocolumn) {
      colParser[key] = PARSERS[field.type] || field.type
    }
  }

  try {
    const data = await csv({ colParser }).fromString(csvString)
    const schemaKeyMap = {}
    Object.keys(schema).forEach(key => (schemaKeyMap[key.toLowerCase()] = key))
    for (let element of data) {
      if (!data) {
        continue
      }
      for (let key of Object.keys(element)) {
        const mappedKey = schemaKeyMap[key.toLowerCase()]
        // isn't a column in the table, remove it
        if (mappedKey == null) {
          delete element[key]
        }
        // casing is different, fix it in row
        else if (key !== mappedKey) {
          element[mappedKey] = element[key]
          delete element[key]
        }
      }
    }
    return data
  } catch (err) {
    console.error(`Error transforming CSV to JSON for data import`, err)
    throw err
  }
}

module.exports = {
  parse,
  transform,
  updateSchema,
}
