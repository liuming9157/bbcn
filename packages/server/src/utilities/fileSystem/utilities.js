const {
  ObjectStore,
  makeSureBucketExists,
  upload,
  deleteFiles,
  streamUpload,
  retrieve,
  retrieveToTmp,
  deleteFolder,
  uploadDirectory,
  downloadTarball,
} = require("@budibase/backend-core/objectStore")

/***********************************
 *              NOTE               *
 * This file purely exists so that *
 *  the object store functionality *
 *   can easily be mocked out of   *
 *  the server without mocking the *
 *      entire core library.       *
 ***********************************/

exports.ObjectStore = ObjectStore
exports.makeSureBucketExists = makeSureBucketExists
exports.upload = upload
exports.streamUpload = streamUpload
exports.retrieve = retrieve
exports.retrieveToTmp = retrieveToTmp
exports.deleteFolder = deleteFolder
exports.uploadDirectory = uploadDirectory
exports.downloadTarball = downloadTarball
exports.deleteFiles = deleteFiles
