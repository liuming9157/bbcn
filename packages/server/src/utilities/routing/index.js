const { createRoutingView } = require("../../db/views/staticViews")
const { ViewNames, getQueryIndex, UNICODE_MAX } = require("../../db/utils")
const { getAppDB } = require("@budibase/backend-core/context")

exports.getRoutingInfo = async () => {
  const db = getAppDB()
  try {
    const allRouting = await db.query(getQueryIndex(ViewNames.ROUTING), {
      startKey: "",
      endKey: UNICODE_MAX,
    })
    return allRouting.rows.map(row => row.value)
  } catch (err) {
    // check if the view doesn't exist, it should for all new instances
    /* istanbul ignore next */
    if (err != null && err.name === "not_found") {
      await createRoutingView()
      return exports.getRoutingInfo()
    } else {
      throw err
    }
  }
}

exports.createRoutingView = createRoutingView
