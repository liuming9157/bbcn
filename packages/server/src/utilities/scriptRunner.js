const fetch = require("node-fetch")
const { VM, VMScript } = require("vm2")
const JS_TIMEOUT_MS = 1000

class ScriptRunner {
  constructor(script, context) {
    const code = `let fn = () => {\n${script}\n}; results.out = fn();`
    this.vm = new VM({
      timeout: JS_TIMEOUT_MS,
    })
    this.results = { out: "" }
    this.vm.setGlobals(context)
    this.vm.setGlobal("fetch", fetch)
    this.vm.setGlobal("results", this.results)
    this.script = new VMScript(code)
  }

  execute() {
    this.vm.run(this.script)
    return this.results.out
  }
}

module.exports = ScriptRunner
