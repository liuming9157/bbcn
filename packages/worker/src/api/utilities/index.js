const { getGlobalDB } = require("@budibase/backend-core/tenancy")
const { getGlobalUserParams } = require("@budibase/backend-core/db")

/**
 * Retrieves all users from the current tenancy.
 */
exports.allUsers = async () => {
  const db = getGlobalDB()
  const response = await db.allDocs(
    getGlobalUserParams(null, {
      include_docs: true,
    })
  )
  return response.rows.map(row => row.doc)
}

/**
 * Gets a user by ID from the global database, based on the current tenancy.
 */
exports.getUser = async userId => {
  const db = getGlobalDB()
  let user
  try {
    user = await db.get(userId)
  } catch (err) {
    // no user found, just return nothing
    user = {}
  }
  if (user) {
    delete user.password
  }
  return user
}
