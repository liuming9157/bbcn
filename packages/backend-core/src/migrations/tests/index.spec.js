require("../../tests/utilities/dbConfig")

const { runMigrations, getMigrationsDoc } = require("../index")
const CouchDB = require("../../db").getCouch()
const {
  StaticDatabases,
} = require("../../db/utils")

Date.now = jest.fn(() => 1487076708000)
let db

describe("migrations", () => {

  const migrationFunction = jest.fn()

  const MIGRATIONS = [{
    type: "global",
    name: "test",
    fn: migrationFunction
  }]

  beforeEach(() => {
    db = new CouchDB(StaticDatabases.GLOBAL.name)
  })

  afterEach(async () => {
    jest.clearAllMocks()
    await db.destroy()
  })

  const migrate = () => {
    return runMigrations(CouchDB, MIGRATIONS)
  }

  it("should run a new migration", async () => {
    await migrate()
    expect(migrationFunction).toHaveBeenCalled()
    const doc = await getMigrationsDoc(db)
    expect(doc.test).toBeDefined()
  })

  it("should match snapshot", async () => {
    await migrate()
    const doc = await getMigrationsDoc(db)
    expect(doc).toMatchSnapshot()
  })

  it("should skip a previously run migration", async () => {
    await migrate()
    const previousMigrationTime = await getMigrationsDoc(db).test
    await migrate()
    const currentMigrationTime = await getMigrationsDoc(db).test
    expect(migrationFunction).toHaveBeenCalledTimes(1)
    expect(currentMigrationTime).toBe(previousMigrationTime)
  })
})