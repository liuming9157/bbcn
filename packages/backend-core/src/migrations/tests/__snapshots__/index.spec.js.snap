// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migrations should match snapshot 1`] = `
Object {
  "_id": "migrations",
  "_rev": "1-6277abc4e3db950221768e5a2618a059",
  "test": 1487076708000,
}
`;
